eth/tracers/logger/logger.go

```diff
@@ -154,7 +154,14 @@ func (l *StructLogger) CaptureStart(env *vm.EVM, from common.Address, to common.
 //
 // CaptureState also tracks SLOAD/SSTORE ops to track storage change.
 func (l *StructLogger) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
-	for ; l.depth > depth-1; l.depth-- {
+	stack := scope.Stack
+	for i := l.depth - (depth - 1); l.depth > depth-1; l.depth, i = l.depth-1, i-1 {
+		if l.current.error == nil {
+			switch stack.Data()[len(stack.Data())-i].Bytes32()[31] {
```

Comment:
It's a bit hard for me to follow the two loop iterators in my head. If I understand correctly, we always have `i == l.depth - (depth - 1)`, right? Maybe instead of tracking `i`, we can calculate
```
stackIndex := len(stack.Data()) - (l.depth - (depth - 1))
```
on each iteration.
This would also allow us to make a bounds check on the variable instead of crashing out-of-bounds if for some reason the three variables `scope.Stack`, `depth` and `l.depth` get out of sync. What do you think?

Reply:
And what do we do when it happens?
NOT crash?

Reply:
Consider making the code clean enough such that the fatal error that is raised will allow some insight into what went wrong. I'm not aware of the context in which this is run (the filename says "logger", but this is not logging code), so if you say a fatal crash is appropriate so be it.

If you could even include a test, that would be amazing.